File: src/main/java/hiiragi283/material/HiiragiBlocks.java
package hiiragi283.material;

import hiiragi283.material.api.block.MaterialBlock;
import hiiragi283.material.api.block.MaterialBlockCasing;
import hiiragi283.material.api.item.HiiragiItemBlock;
import hiiragi283.material.api.registry.HiiragiEntry;
import hiiragi283.material.api.shape.HiiragiShapes;
import hiiragi283.material.block.BlockModuleInstaller;
import hiiragi283.material.block.BlockTestMachine;
import hiiragi283.material.config.RMConfig;
import hiiragi283.material.util.CraftingBuilder;
import hiiragi283.material.util.HiiragiUtil;
import net.minecraft.block.Block;
import net.minecraft.client.renderer.color.BlockColors;
import net.minecraft.client.renderer.color.ItemColors;
import net.minecraftforge.fml.relauncher.Side;
import net.minecraftforge.fml.relauncher.SideOnly;
import net.minecraftforge.registries.IForgeRegistry;
import org.jetbrains.annotations.Nullable;

import java.util.ArrayList;
import java.util.List;
import java.util.Objects;
import java.util.stream.Collectors;

public class HiiragiBlocks implements HiiragiEntry.BLOCK {

    public static HiiragiBlocks INSTANCE = new HiiragiBlocks();

    private HiiragiBlocks() {
    }

    private static final List<BLOCK> ENTRIES = new ArrayList<>();

    public static List<HiiragiEntry.ITEM> getItemBlocks() {
        return ENTRIES.stream()
                .map(HiiragiEntry.BLOCK::getItemBlock)
                .filter(Objects::nonNull)
                .collect(Collectors.toList());
    }

    //    Blocks    //

    public static final MaterialBlock MATERIAL_BLOCK = MaterialBlock.create(
            HiiragiShapes.BLOCK,
            (block, material) -> {
                if (material.isMetal()) {
                    new CraftingBuilder(block.asItemStack(material))
                            .setPattern("AAA", "AAA", "AAA")
                            .setIngredient('A', HiiragiShapes.INGOT.getOreDict(material))
                            .build();
                } else if (material.isGem()) {
                    new CraftingBuilder(block.asItemStack(material))
                            .setPattern("AAA", "AAA", "AAA")
                            .setIngredient('A', HiiragiShapes.GEM.getOreDict(material))
                            .build();
                }
            }
    );

    public static final MaterialBlock MATERIAL_CASING = new MaterialBlockCasing();

    public static final MaterialBlock MATERIAL_FRAME = MaterialBlock.create(
            HiiragiShapes.FRAME,
            (block, material) -> new CraftingBuilder(block.asItemStack(material, 2))
                    .setPattern("AAA", "A A", "AAA")
                    .setIngredient('A', HiiragiShapes.STICK.getOreDict(material))
                    .build()
    );

    public static final BlockTestMachine TEST_MACHINE = new BlockTestMachine();

    public static final BlockModuleInstaller MODULE_INSTALLER = new BlockModuleInstaller();

    //    Registry    //

    public static void init() {
        RagiMaterials.LOGGER.info("HiiragiBlocks has been initialized!");
        if (RMConfig.EXPERIMENTAL.enableMetaTileBlock) {
            ENTRIES.add(MATERIAL_BLOCK);
            ENTRIES.add(MATERIAL_CASING);
            ENTRIES.add(MATERIAL_FRAME);
            MATERIAL_BLOCK.registerTileEntity();
        }
        if (HiiragiUtil.isDeobf()) {
            ENTRIES.add(TEST_MACHINE);
            ENTRIES.add(MODULE_INSTALLER);
            TEST_MACHINE.registerTileEntity();
            MODULE_INSTALLER.registerTileEntity();
        }
    }

    @Override
    public @Nullable HiiragiItemBlock getItemBlock() {
        return null;
    }

    @Override
    public void register(IForgeRegistry<Block> registry) {
        ENTRIES.forEach(block -> registry.register(block.getObject()));
    }

    @Override
    public void registerOreDict() {
        ENTRIES.forEach(HiiragiEntry::registerOreDict);
    }

    @Override
    public void registerRecipe() {
        ENTRIES.forEach(HiiragiEntry::registerRecipe);
    }

    @Override
    @SideOnly(Side.CLIENT)
    public void registerBlockColor(BlockColors blockColors) {
        ENTRIES.forEach(block -> block.registerBlockColor(blockColors));
    }

    @Override
    @SideOnly(Side.CLIENT)
    public void registerItemColor(ItemColors itemColors) {
        ENTRIES.forEach(block -> block.registerItemColor(itemColors));
    }

    @Override
    @SideOnly(Side.CLIENT)
    public void registerModel() {
        ENTRIES.forEach(HiiragiEntry::registerModel);
    }

}

File: src/main/kotlin/hiiragi283/material/HiiragiBlocks.kt
package hiiragi283.material

import hiiragi283.material.api.registry.HiiragiRegistries
import hiiragi283.material.api.tile.MaterialTileEntity
import hiiragi283.material.block.*
import hiiragi283.material.config.HiiragiConfigs
import hiiragi283.material.tile.*
import hiiragi283.material.util.hiiragiLocation
import hiiragi283.material.util.isDeobf
import net.minecraft.tileentity.TileEntity
import net.minecraftforge.energy.CapabilityEnergy
import net.minecraftforge.fluids.capability.CapabilityFluidHandler
import net.minecraftforge.fml.common.registry.GameRegistry
import net.minecraftforge.items.CapabilityItemHandler

object HiiragiBlocks {

    //    Material    //

    @JvmField
    val MATERIAL_BLOCK =
        HiiragiRegistries.BLOCK.registerOptional(MaterialBlockStorage)
        { HiiragiConfigs.EXPERIMENTAL.enableMetaTileBlock }

    @JvmField
    val MATERIAL_CASING =
        HiiragiRegistries.BLOCK.registerOptional(MaterialBlockCasing)
        { HiiragiConfigs.EXPERIMENTAL.enableMetaTileBlock }


    //    Machine    //

    @JvmField
    val MACHINE_EXTENDER = HiiragiRegistries.BLOCK.register(BlockMachineExtender)

    @JvmField
    val MACHINE_WORKBENCH = HiiragiRegistries.BLOCK.register(BlockMachineWorkbench)

    //    Common    //

    @JvmField
    val CAPABILITY_RAIL = HiiragiRegistries.BLOCK.register(BlockCapabilityRail)

    @JvmField
    val ITEM_PIPE = HiiragiRegistries.BLOCK.register(
        BlockTransferPipe("item") { listOf(CapabilityItemHandler.ITEM_HANDLER_CAPABILITY) }
    )

    @JvmField
    val FLUID_PIPE = HiiragiRegistries.BLOCK.register(
        BlockTransferPipe("fluid") { listOf(CapabilityFluidHandler.FLUID_HANDLER_CAPABILITY) }
    )

    @JvmField
    val ENERGY_PIPE = HiiragiRegistries.BLOCK.register(
        BlockTransferPipe("energy") { listOf(CapabilityEnergy.ENERGY) }
    )

    @JvmField
    val ALL_PIPE = HiiragiRegistries.BLOCK.register(
        BlockTransferPipe("all") {
            listOf(
                CapabilityItemHandler.ITEM_HANDLER_CAPABILITY,
                CapabilityFluidHandler.FLUID_HANDLER_CAPABILITY,
                CapabilityEnergy.ENERGY
            )
        }
    )

    @JvmField
    val ITEM_STATION = HiiragiRegistries.BLOCK.register(
        BlockTransferStation("item") { TileTransferStationItem() }
    )

    @JvmField
    val FLUID_STATION = HiiragiRegistries.BLOCK.register(
        BlockTransferStation("fluid") { TileTransferStationFluid() }
    )

    @JvmField
    val ENERGY_STATION = HiiragiRegistries.BLOCK.register(
        BlockTransferStation("energy") { TileTransferStationEnergy() }
    )

    //    TEST    //

    @JvmField
    val CHUNK_LOADER = HiiragiRegistries.BLOCK.registerOptional(BlockTestChunkLoader) { isDeobf() }

    @JvmField
    val TEST_CORE = HiiragiRegistries.BLOCK.registerOptional(BlockTestMultiblock) { isDeobf() }

    init {
        registerTileEntity(MaterialTileEntity::class.java, "material")
        registerTileEntity(TileEntityCapabilityRail::class.java, "capability_rail")
        registerTileEntity(TileEntityMachineExtender::class.java, "machine_extender")
        registerTileEntity(TileEntityModuleMachine::class.java, "module_machine")
        registerTileEntity(TileTransferStationEnergy::class.java, "energy_station")
        registerTileEntity(TileTransferStationFluid::class.java, "fluid_station")
        registerTileEntity(TileTransferStationItem::class.java, "item_station")
    }

    private fun <T : TileEntity> registerTileEntity(clazz: Class<T>, name: String) {
        GameRegistry.registerTileEntity(clazz, hiiragiLocation(name))
    }

}